<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.11"/>
<title>Ymir: ymir::ProbabilisticAssemblingModel Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { init_search(); });
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">Ymir
   &#160;<span id="projectnumber">.9</span>
   </div>
   <div id="projectbrief">Fast\C++toolforcomputationofassemblingprobabilities,statisticalinferenceofassemblingstatisticalmodelandgenerationofartificialsequencesofT-cellreceptorsdata.</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.11 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="annotated.html"><span>Class&#160;List</span></a></li>
      <li><a href="classes.html"><span>Class&#160;Index</span></a></li>
      <li><a href="hierarchy.html"><span>Class&#160;Hierarchy</span></a></li>
      <li><a href="functions.html"><span>Class&#160;Members</span></a></li>
    </ul>
  </div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><b>ymir</b></li><li class="navelem"><a class="el" href="classymir_1_1_probabilistic_assembling_model.html">ProbabilisticAssemblingModel</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pro-methods">Protected Member Functions</a> &#124;
<a href="#pro-attribs">Protected Attributes</a> &#124;
<a href="classymir_1_1_probabilistic_assembling_model-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">ymir::ProbabilisticAssemblingModel Class Reference</div>  </div>
</div><!--header-->
<div class="contents">
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:aca2e51265f946379d3f58296d400445c"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classymir_1_1_probabilistic_assembling_model.html#aca2e51265f946379d3f58296d400445c">ProbabilisticAssemblingModel</a> (const string &amp;folderpath, ModelBehaviour behav=PREDEFINED)</td></tr>
<tr class="memdesc:aca2e51265f946379d3f58296d400445c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Model constructor from the given folder with a model's parameters.  <a href="#aca2e51265f946379d3f58296d400445c">More...</a><br /></td></tr>
<tr class="separator:aca2e51265f946379d3f58296d400445c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6b3ee8154be898a7c47fa6ca9a7764fc"><td class="memItemLeft" align="right" valign="top">vector&lt; prob_t &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classymir_1_1_probabilistic_assembling_model.html#a6b3ee8154be898a7c47fa6ca9a7764fc">computeFullProbabilities</a> (const <a class="el" href="classymir_1_1_cloneset_view.html">ClonesetView</a> &amp;repertoire, ErrorMode error_mode, SequenceType sequence_type=NUCLEOTIDE, MAAGComputeProbAction action=SUM_PROBABILITY) const </td></tr>
<tr class="memdesc:a6b3ee8154be898a7c47fa6ca9a7764fc"><td class="mdescLeft">&#160;</td><td class="mdescRight">Compute full assembling probabilities of clonotypes from the given repertoire. If some clonotypes have few alignments of some gene segments, than full probability of all combinations of gene segments will be computed and either summed or the max full probability will be chosen.  <a href="#a6b3ee8154be898a7c47fa6ca9a7764fc">More...</a><br /></td></tr>
<tr class="separator:a6b3ee8154be898a7c47fa6ca9a7764fc"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a839b331ab03ececaea940d2720e1e337"><td class="memItemLeft" align="right" valign="top">MAAGRepertoire&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classymir_1_1_probabilistic_assembling_model.html#a839b331ab03ececaea940d2720e1e337">buildGraphs</a> (const <a class="el" href="classymir_1_1_cloneset_view.html">ClonesetView</a> &amp;repertoire, MetadataMode save_metadata, ErrorMode error_mode, SequenceType sequence_type=NUCLEOTIDE, bool verbose=true) const </td></tr>
<tr class="memdesc:a839b331ab03ececaea940d2720e1e337"><td class="mdescLeft">&#160;</td><td class="mdescRight">Build a set of MAAGs from the given cloneset.  <a href="#a839b331ab03ececaea940d2720e1e337">More...</a><br /></td></tr>
<tr class="separator:a839b331ab03ececaea940d2720e1e337"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aaeb4712649ef2b3e79d778da866e8868"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aaeb4712649ef2b3e79d778da866e8868"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classymir_1_1_probabilistic_assembling_model.html#aaeb4712649ef2b3e79d778da866e8868">updateEventProbabilities</a> (MAAGRepertoire *repertoire, bool verbose=true)</td></tr>
<tr class="memdesc:aaeb4712649ef2b3e79d778da866e8868"><td class="mdescLeft">&#160;</td><td class="mdescRight">Update event probabilities in the given <a class="el" href="classymir_1_1_m_a_a_g.html" title="Multi-Alignment Assembly Graph - basic class for representing all possible generation scenarios of a ...">MAAG</a> repertoire with new ones. <br /></td></tr>
<tr class="separator:aaeb4712649ef2b3e79d778da866e8868"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad4ba559c350c0012a19857a0024b1a82"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classymir_1_1_cloneset.html">Cloneset</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classymir_1_1_probabilistic_assembling_model.html#ad4ba559c350c0012a19857a0024b1a82">generateSequences</a> (size_t count=1, bool verbose=true) const </td></tr>
<tr class="memdesc:ad4ba559c350c0012a19857a0024b1a82"><td class="mdescLeft">&#160;</td><td class="mdescRight">Generate artificial repertoire of sequences using this model.  <a href="#ad4ba559c350c0012a19857a0024b1a82">More...</a><br /></td></tr>
<tr class="separator:ad4ba559c350c0012a19857a0024b1a82"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a10abac39f71dd89da74f1a83dfdd1426"><td class="memItemLeft" align="right" valign="top">const <a class="el" href="classymir_1_1_v_d_j_recombination_genes.html">VDJRecombinationGenes</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classymir_1_1_probabilistic_assembling_model.html#a10abac39f71dd89da74f1a83dfdd1426">gene_segments</a> () const </td></tr>
<tr class="memdesc:a10abac39f71dd89da74f1a83dfdd1426"><td class="mdescLeft">&#160;</td><td class="mdescRight">Access to the gene segments table.  <a href="#a10abac39f71dd89da74f1a83dfdd1426">More...</a><br /></td></tr>
<tr class="separator:a10abac39f71dd89da74f1a83dfdd1426"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a89683ef69f2d01bd4911c369c9257226"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a89683ef69f2d01bd4911c369c9257226"></a>
const <a class="el" href="classymir_1_1_model_parameter_vector.html">ModelParameterVector</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classymir_1_1_probabilistic_assembling_model.html#a89683ef69f2d01bd4911c369c9257226">event_probabilities</a> () const </td></tr>
<tr class="memdesc:a89683ef69f2d01bd4911c369c9257226"><td class="mdescLeft">&#160;</td><td class="mdescRight">Access to vector of probabilities. <br /></td></tr>
<tr class="separator:a89683ef69f2d01bd4911c369c9257226"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a446d4efbdc5b2b3879e5fc5c24079c36"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classymir_1_1_probabilistic_assembling_model.html#a446d4efbdc5b2b3879e5fc5c24079c36">updateModelParameterVector</a> (const <a class="el" href="classymir_1_1_model_parameter_vector.html">ModelParameterVector</a> &amp;vec)</td></tr>
<tr class="memdesc:a446d4efbdc5b2b3879e5fc5c24079c36"><td class="mdescLeft">&#160;</td><td class="mdescRight">Set a new event probabilities vector to this model.  <a href="#a446d4efbdc5b2b3879e5fc5c24079c36">More...</a><br /></td></tr>
<tr class="separator:a446d4efbdc5b2b3879e5fc5c24079c36"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a866cc2edf25032a14f364b4c345e851c"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a866cc2edf25032a14f364b4c345e851c"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classymir_1_1_probabilistic_assembling_model.html#a866cc2edf25032a14f364b4c345e851c">updateGeneUsage</a> (const <a class="el" href="classymir_1_1_cloneset_view.html">ClonesetView</a> &amp;cloneset)</td></tr>
<tr class="memdesc:a866cc2edf25032a14f364b4c345e851c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Given the cloneset, compute its gene usage on out-of-frames and update the event probability vector with new gene probabilities. <br /></td></tr>
<tr class="separator:a866cc2edf25032a14f364b4c345e851c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af1707606ef8f41ea4068d93e301a389b"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classymir_1_1_probabilistic_assembling_model.html#af1707606ef8f41ea4068d93e301a389b">save</a> (const string &amp;folderpath) const </td></tr>
<tr class="memdesc:af1707606ef8f41ea4068d93e301a389b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Save the model to the given folder.  <a href="#af1707606ef8f41ea4068d93e301a389b">More...</a><br /></td></tr>
<tr class="separator:af1707606ef8f41ea4068d93e301a389b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad7b7b05a1d38480bb41de8d5c93c3847"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classymir_1_1_probabilistic_assembling_model.html#ad7b7b05a1d38480bb41de8d5c93c3847">status</a> () const </td></tr>
<tr class="memdesc:ad7b7b05a1d38480bb41de8d5c93c3847"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return if the model is in the ready-to-work state.  <a href="#ad7b7b05a1d38480bb41de8d5c93c3847">More...</a><br /></td></tr>
<tr class="separator:ad7b7b05a1d38480bb41de8d5c93c3847"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a687cd3794ee0001557c307a57e2ee50c"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a687cd3794ee0001557c307a57e2ee50c"></a>
Recombination&#160;</td><td class="memItemRight" valign="bottom"><b>recombination</b> () const </td></tr>
<tr class="separator:a687cd3794ee0001557c307a57e2ee50c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa9af3593bc6e7294befb095dba01c8c7"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aa9af3593bc6e7294befb095dba01c8c7"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><b>name</b> () const </td></tr>
<tr class="separator:aa9af3593bc6e7294befb095dba01c8c7"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pro-methods"></a>
Protected Member Functions</h2></td></tr>
<tr class="memitem:ab2e394e13dd19e96d24c885b76d45659"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ab2e394e13dd19e96d24c885b76d45659"></a>
&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classymir_1_1_probabilistic_assembling_model.html#ab2e394e13dd19e96d24c885b76d45659">ProbabilisticAssemblingModel</a> ()</td></tr>
<tr class="memdesc:ab2e394e13dd19e96d24c885b76d45659"><td class="mdescLeft">&#160;</td><td class="mdescRight">Private default constructor. <br /></td></tr>
<tr class="separator:ab2e394e13dd19e96d24c885b76d45659"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7bea84b29848896f451987100af219ef"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a7bea84b29848896f451987100af219ef"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classymir_1_1_probabilistic_assembling_model.html#a7bea84b29848896f451987100af219ef">parseModelConfig</a> (const string &amp;folderpath)</td></tr>
<tr class="memdesc:a7bea84b29848896f451987100af219ef"><td class="mdescLeft">&#160;</td><td class="mdescRight">Parse JSON file with model parameters. <br /></td></tr>
<tr class="separator:a7bea84b29848896f451987100af219ef"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa452c5399f82c563270a78108dd23a85"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aa452c5399f82c563270a78108dd23a85"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classymir_1_1_probabilistic_assembling_model.html#aa452c5399f82c563270a78108dd23a85">make_builder</a> ()</td></tr>
<tr class="memdesc:aa452c5399f82c563270a78108dd23a85"><td class="mdescLeft">&#160;</td><td class="mdescRight">Make <a class="el" href="classymir_1_1_m_a_a_g_builder.html">MAAGBuilder</a> with this model's parameters (event probabilities and gene segments). <br /></td></tr>
<tr class="separator:aa452c5399f82c563270a78108dd23a85"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abfdbc1b6bd8adeec67518f9fb46694fc"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="abfdbc1b6bd8adeec67518f9fb46694fc"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classymir_1_1_probabilistic_assembling_model.html#abfdbc1b6bd8adeec67518f9fb46694fc">make_assembler</a> ()</td></tr>
<tr class="memdesc:abfdbc1b6bd8adeec67518f9fb46694fc"><td class="mdescLeft">&#160;</td><td class="mdescRight">Make <a class="el" href="classymir_1_1_clonotype_assembler.html">ClonotypeAssembler</a> with this model's parameters (event probabilities and gene segments). <br /></td></tr>
<tr class="separator:abfdbc1b6bd8adeec67518f9fb46694fc"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a523344466f9e646e919975e39e484be0"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a523344466f9e646e919975e39e484be0"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><b>save_vj</b> (const string &amp;folderpath) const </td></tr>
<tr class="separator:a523344466f9e646e919975e39e484be0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:adecae84038ae8b07c3b864195007d5e7"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="adecae84038ae8b07c3b864195007d5e7"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><b>save_vdj</b> (const string &amp;folderpath) const </td></tr>
<tr class="separator:adecae84038ae8b07c3b864195007d5e7"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pro-attribs"></a>
Protected Attributes</h2></td></tr>
<tr class="memitem:a6ac56b5ba064e2facd1fc46bd87ada25"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a6ac56b5ba064e2facd1fc46bd87ada25"></a>
bool&#160;</td><td class="memItemRight" valign="bottom"><b>_status</b></td></tr>
<tr class="separator:a6ac56b5ba064e2facd1fc46bd87ada25"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5306da6d5111a03487c95d6bee4de575"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a5306da6d5111a03487c95d6bee4de575"></a>
bool&#160;</td><td class="memItemRight" valign="bottom"><b>_verbose</b></td></tr>
<tr class="separator:a5306da6d5111a03487c95d6bee4de575"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2002101dce08395b886f147114adc555"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a2002101dce08395b886f147114adc555"></a>
ModelBehaviour&#160;</td><td class="memItemRight" valign="bottom"><b>_behaviour</b></td></tr>
<tr class="separator:a2002101dce08395b886f147114adc555"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac163fbdd5d618f52d130f77832f1606d"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ac163fbdd5d618f52d130f77832f1606d"></a>
Json::Value&#160;</td><td class="memItemRight" valign="bottom"><b>_config</b></td></tr>
<tr class="separator:ac163fbdd5d618f52d130f77832f1606d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a23333054687b158494f4cb6bdc39dcc7"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a23333054687b158494f4cb6bdc39dcc7"></a>
Recombination&#160;</td><td class="memItemRight" valign="bottom"><b>_recomb</b></td></tr>
<tr class="separator:a23333054687b158494f4cb6bdc39dcc7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a42dcb1e9477e23e789b019cbc61fb471"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a42dcb1e9477e23e789b019cbc61fb471"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><b>_model_path</b></td></tr>
<tr class="separator:a42dcb1e9477e23e789b019cbc61fb471"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9b5dd75c26cdd4e9ad6e889a6e2c8882"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a9b5dd75c26cdd4e9ad6e889a6e2c8882"></a>
unique_ptr&lt; <a class="el" href="classymir_1_1_model_parser.html">ModelParser</a> &gt;&#160;</td><td class="memItemRight" valign="bottom"><b>_parser</b></td></tr>
<tr class="separator:a9b5dd75c26cdd4e9ad6e889a6e2c8882"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abe3d80c35fe65ded3dbb6afdec06af2f"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="abe3d80c35fe65ded3dbb6afdec06af2f"></a>
unique_ptr&lt; <a class="el" href="classymir_1_1_v_d_j_recombination_genes.html">VDJRecombinationGenes</a> &gt;&#160;</td><td class="memItemRight" valign="bottom"><b>_genes</b></td></tr>
<tr class="separator:abe3d80c35fe65ded3dbb6afdec06af2f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6636d867fa63eea22905f44ee1a603be"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a6636d867fa63eea22905f44ee1a603be"></a>
unique_ptr&lt; <a class="el" href="classymir_1_1_model_parameter_vector.html">ModelParameterVector</a> &gt;&#160;</td><td class="memItemRight" valign="bottom"><b>_param_vec</b></td></tr>
<tr class="separator:a6636d867fa63eea22905f44ee1a603be"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:adfa38166e6ea3a1ae3af889a4fa382c1"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="adfa38166e6ea3a1ae3af889a4fa382c1"></a>
unique_ptr&lt; <a class="el" href="classymir_1_1_m_a_a_g_builder.html">MAAGBuilder</a> &gt;&#160;</td><td class="memItemRight" valign="bottom"><b>_builder</b></td></tr>
<tr class="separator:adfa38166e6ea3a1ae3af889a4fa382c1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9de68a1a5199975c9e8cc2f875f03971"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a9de68a1a5199975c9e8cc2f875f03971"></a>
unique_ptr&lt; <a class="el" href="classymir_1_1_clonotype_assembler.html">ClonotypeAssembler</a> &gt;&#160;</td><td class="memItemRight" valign="bottom"><b>_generator</b></td></tr>
<tr class="separator:a9de68a1a5199975c9e8cc2f875f03971"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af9c1eebe38d30211b71dafa6b64e2193"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="af9c1eebe38d30211b71dafa6b64e2193"></a>
seq_len_t&#160;</td><td class="memItemRight" valign="bottom"><b>_min_D_len</b></td></tr>
<tr class="separator:af9c1eebe38d30211b71dafa6b64e2193"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<h2 class="groupheader">Constructor &amp; Destructor Documentation</h2>
<a class="anchor" id="aca2e51265f946379d3f58296d400445c"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">ymir::ProbabilisticAssemblingModel::ProbabilisticAssemblingModel </td>
          <td>(</td>
          <td class="paramtype">const string &amp;&#160;</td>
          <td class="paramname"><em>folderpath</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">ModelBehaviour&#160;</td>
          <td class="paramname"><em>behav</em> = <code>PREDEFINED</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Model constructor from the given folder with a model's parameters. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">folderpath</td><td>Path to a folder with a model's parameters. </td></tr>
    <tr><td class="paramname">behav</td><td>Model's behaviour. PREDEFINED means that each probability will stay constant for each input data unless the model probabilities vector updated with new parameters (either all parameters or gene usage only). EMPTY means that the model will load only gene segments sequences and generate vector of uniformly distributed marginal probabilities. This model could be used as an input model for the statistical inference of the model's parameters. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a class="anchor" id="a839b331ab03ececaea940d2720e1e337"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">MAAGRepertoire ymir::ProbabilisticAssemblingModel::buildGraphs </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classymir_1_1_cloneset_view.html">ClonesetView</a> &amp;&#160;</td>
          <td class="paramname"><em>repertoire</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">MetadataMode&#160;</td>
          <td class="paramname"><em>save_metadata</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">ErrorMode&#160;</td>
          <td class="paramname"><em>error_mode</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">SequenceType&#160;</td>
          <td class="paramname"><em>sequence_type</em> = <code>NUCLEOTIDE</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>verbose</em> = <code>true</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Build a set of MAAGs from the given cloneset. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">repertoire</td><td><a class="el" href="classymir_1_1_cloneset.html">Cloneset</a>. </td></tr>
    <tr><td class="paramname">save_metadatae</td><td>If true than build full <a class="el" href="classymir_1_1_m_a_a_g.html" title="Multi-Alignment Assembly Graph - basic class for representing all possible generation scenarios of a ...">MAAG</a> (with additional matrices for event indices). </td></tr>
    <tr><td class="paramname">aminoacid</td><td>If true than build graphs of aminoacid CDR3 sequences, traslating the given repertoire first.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Set of MAAGs. </dd></dl>

</div>
</div>
<a class="anchor" id="a6b3ee8154be898a7c47fa6ca9a7764fc"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">vector&lt;prob_t&gt; ymir::ProbabilisticAssemblingModel::computeFullProbabilities </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classymir_1_1_cloneset_view.html">ClonesetView</a> &amp;&#160;</td>
          <td class="paramname"><em>repertoire</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">ErrorMode&#160;</td>
          <td class="paramname"><em>error_mode</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">SequenceType&#160;</td>
          <td class="paramname"><em>sequence_type</em> = <code>NUCLEOTIDE</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">MAAGComputeProbAction&#160;</td>
          <td class="paramname"><em>action</em> = <code>SUM_PROBABILITY</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Compute full assembling probabilities of clonotypes from the given repertoire. If some clonotypes have few alignments of some gene segments, than full probability of all combinations of gene segments will be computed and either summed or the max full probability will be chosen. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">repertoire</td><td>Repertoire with alignments and event indices. </td></tr>
    <tr><td class="paramname">aminoacid</td><td>What probabilities will be computed - of nucleotide or amino acid sequences. If this parameter is false and some clone will have amino acid sequence, than error message will be generated to the output and probability of this clone will set to zero.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Vector of full assembling probabilities. </dd></dl>

</div>
</div>
<a class="anchor" id="a10abac39f71dd89da74f1a83dfdd1426"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">const <a class="el" href="classymir_1_1_v_d_j_recombination_genes.html">VDJRecombinationGenes</a>&amp; ymir::ProbabilisticAssemblingModel::gene_segments </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Access to the gene segments table. </p>
<dl class="section return"><dt>Returns</dt><dd>Const reference to the gene segments table. </dd></dl>

</div>
</div>
<a class="anchor" id="ad4ba559c350c0012a19857a0024b1a82"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classymir_1_1_cloneset.html">Cloneset</a> ymir::ProbabilisticAssemblingModel::generateSequences </td>
          <td>(</td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>count</em> = <code>1</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>verbose</em> = <code>true</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Generate artificial repertoire of sequences using this model. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">count</td><td>Number of generated sequences. </td></tr>
    <tr><td class="paramname">merge</td><td>If true than return only unique clonotypes with counts of occurrences for each clonotype.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Artificial repertoire. </dd></dl>

</div>
</div>
<a class="anchor" id="af1707606ef8f41ea4068d93e301a389b"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">bool ymir::ProbabilisticAssemblingModel::save </td>
          <td>(</td>
          <td class="paramtype">const string &amp;&#160;</td>
          <td class="paramname"><em>folderpath</em></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Save the model to the given folder. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">folderpath</td><td>Path to the model folder.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>True if all is ok. </dd></dl>

</div>
</div>
<a class="anchor" id="ad7b7b05a1d38480bb41de8d5c93c3847"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">bool ymir::ProbabilisticAssemblingModel::status </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Return if the model is in the ready-to-work state. </p>
<dl class="section return"><dt>Returns</dt><dd>True or false. </dd></dl>

</div>
</div>
<a class="anchor" id="a446d4efbdc5b2b3879e5fc5c24079c36"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void ymir::ProbabilisticAssemblingModel::updateModelParameterVector </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classymir_1_1_model_parameter_vector.html">ModelParameterVector</a> &amp;&#160;</td>
          <td class="paramname"><em>vec</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Set a new event probabilities vector to this model. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">vec</td><td>Vector with new event probabilities. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>Ymir/src/<a class="el" href="probabilisticassemblingmodel_8h_source.html">probabilisticassemblingmodel.h</a></li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated on Thu Mar 24 2016 23:59:15 for Ymir by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.11
</small></address>
</body>
</html>
